### ISSUE TITLE
Simplified Jira Integrations

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#2773]
Integrating projects is an unnecessary step, so I'd like a simpler Jira integration.

Here's the states:

CARDS
- The icon has a "Export task" tooltip. 
- If the task assignee has 0 integrations & i'm the assignee, I see a publish icon. Clicking it shows a menu with 2 options "Connect GitHUb" and "Connect Jira". Clicking one of them will open an oauth modal for that provider. After they approve the oauth modal, they will have 1 integration. If i'm not the assignee, there's a message saying "{Assignee} has no integrations set up"
- if the assignee has 1 integration, the publish icon is replaced with the icon of the integration. Clicking it shows a menu of <= 3 Recent projects | Connect Other | Search | Search Results. If there are no recent tasks, the search results are autopopulated. If i click a recent  project, it gets exported, if i click Connect Other, it opens an oauth window. if i type in the search field, a typeahead type filter is applied to the list of search results.
- if the assignee has 2+ integrations, i see a publish icon. clicking it shows the same menu. The search field searches across both integrations.    

- to be clear, if i change the assignee, the integration icon could change & the list of integrations to select from changes. this means anyone can use anyone else's token to assign them tasks. this avoids creating a 1-to-1 list of parabol user to integration user. wins all around!  

INTEGRATION GRID VIEW

If I'm not integrated to Jira:
- Primary action: "Connect" button which opens an oauth2 window to connect to a single jira site
- can connect from card menu or integration list/grid view
- there is an affordance for an error message from trying to integrate (similar to login)

If I am integrated:
- Primary action: add another site
- The Confluence integration is visible & I can integrate with it (out of scope)
- The names/avatars of the sites that I'm integrated with are shown
- Secondary actions:
  - remove entire integration (hidden behind a menu)
  - remove/blacklist a specific site (out of scope, but maybe for the future) probably click on the site avatar to open a menu? Plan for the affordance, but let's not build it.

**Comment by ackernaut** [comment]
In Jira an issue has a reporter (aka creator) and can be unassigned. I’m wondering:

1) If I’m connected, I can report an issue (exporting the task to Jira). My user is the reporter
2) If the Parabol task owner is connected, we could opt to assign the Jira issue to them, but I wonder if this is bad UX because this task probably needs to go into the backlog and be parsed out in a future sprint

Out-of-band UX: if the person creating the Parabol task doesn’t live in Jira at all, then they probably shouldn’t be pushing things out? If they do have a user but haven’t connected, well they should connect. Connected users can always push things out at their own discretion.

At any rate we need to ship something and learn, see what pains users have

PS. Somewhere we need to capture the requirement that export Parabol tasks are created as issues in the backlog (as the default UX) of the Jira project.

**Comment by mattkrick** [comment]
i think its fair to say that the reporter is the person creating the issue. Since we don't have a way to have a Task without an assignee, we've got a hard constraint to maintain 1 assignee in our system.

Here's the proposed criteria:
- The task creator is the reporter. If the task creator is not on Jira, then the task assignee is the reporter.
- The assignee is the assignee

**Comment by ackernaut** [comment]
I think it would be good to ask some more folks if 1) they expect cards coming from Parabol to be in the backlog 2) they except them to be assigned or unassigned because of any flow around sprint planning etc. — who do we talk to @jordanh ?

**Comment by mattkrick** [comment]
that's good, i'll pause on integrations for now @jordanh blocked until we get this resolved

**Comment by jordanh** [comment]
Unblocking you @mattkrick!

Assertion: in the median agile workflow on Jira, assigning isn't done until sprint planning. So:

- The Parabol Task creator is the Jira Issue reporter
- The Jira Issue does unassigned (for now)

This will mirror our existing behavior with GitHub, and give us a solid footing at which to iterate from. Just wait until we have card backs that differ based on the integration!

**Comment by mattkrick** [comment]
@jordanh how does a non-jira user create a jira task? that seems like a core function necessary for cross functional teams

**Comment by jordanh** [comment]
For starters, they don't. All of the teams I've spoken to so far are dev teams, and everybody has a Jira account. So, if you're not integrated you first have to assign the card to somebody who is.

**Comment by mattkrick** [comment]
That makes things easy, but if we build it this way now (possible integrations populated by viewer tokens) , we won't be able to switch later without a TON of rework (for all integrations, not just jira).

For context, here's the same problem but for the github integration: https://github.com/ParabolInc/action/issues/2513

 IMO, the ability to assign someone work without being in their tool of choice is _the_ core feature of the parabol.

**Comment by jordanh** [comment]
What do you need?

**Comment by mattkrick** [comment]
a design that lets parabol user A create a task for user B in system C, given that user B has access to system C & user A does not.

**Comment by jordanh** [comment]
Thank you for the clarification. That's helpful.

Given:

> a design that lets parabol user A create a task for user B in system C, given that user B has access to system C & user A does not.

I can't imagine a case where you'd want user A to use just anybody's credentials—I imagine that each project _may_ have 1 user who voluntarily offers their credentials up for use by the rest of the team.

The obvious question, which is what I believe you're pointing out here, would be where to put this. I'd like to stuff it in the team settings integration configuration modal itself (keeping it off the card face menus) because I think it's advanced baseball. We'd need to spend a design cycle on it.

Is there a way we can decouple the tech development from the design on this? I'd like to ship a beta of what people are asking for (we've got a dozen teams waiting) and then iterate and get the UI designed and implemented to configure delegation.

Is that possible?

**Comment by mattkrick** [comment]
no team settings configs needed & folks don't need to offer up their credentials.

Here's an example:

- Matt belongs to Project Action in Jira & is integrated
- Jordan belongs to Project Action and Project Srikeforce in Jira & is integrated with both
- Taya is integrated with neither
- Jordan creates a task, assigns Matt, and clicks publish
- The only option Jordan sees is Action
- Jordan changes the assignee to himself & now sees 2 options: Strikeforce & Action
- Taya creates a Task & assigns it to Matt, she clicks publish & sees the option Action
- Taya changes the assignee to herself & has 0 options
- Taya changes the assignee to Jordan & has Strikeforce & Action

In the above example, the possible projects is populated by the assignee's credentials, not the viewer's. That means there's no possibility of bad perms & there's no lookup table required to connect parabol users to jira users since we simple filter out the folks who haven't integrated.

I think the root cause is that parabol wasn't build to support unassigned tasks & now we're trying to do that. Either we need to support that, or we need to assign users in Jira upon creation. The alternative is an implicit rule like "the assignee in parabol is equivalent to the reporter in jira until jira has an assignee, and then it means assignee". i'd really like to avoid that kinda confusing logic

**Comment by jordanh** [comment]
Thank you for writing out that use case. It's very helpful, and, very confirming that we're talking about the same things.

For Jira, GitHub, and likely others I think the card front avatar really mean "reported by" when used in an integration context because that is how the credentials are used. In the future (where everything is better) we can iterate the design to have clearer reported by / assigned to affordances. Let's let it be implicit for now—learn a bit more.

I want to avoid being blocked. If you want to talk, let me know, I am around!

**Comment by mattkrick** [comment]
> For Jira, GitHub, and likely others I think the card front avatar really mean "reported by" when used in an integration context because that is how the credentials are used.

Using that logic for the example above: if Jordan assigns a card to Matt & then clicks publish to Strikeforce, then that causes an error because Matt does not have permission. More importantly, if Jordan assigns a card to Matt & then clicks publish to Action, _then the parabol assignee would revert back to Jordan since he created it_. Could you imagine if you hit the publish button & when you did the face changed to yours instead of the person you wrote the card for? 10/10 would call that a bug.

Here are the options I see:
- we replace the user-supplied "assignee" with the viewer, as described above (this feels like a bug & not safe to try)
- same as above, but we explain that instead of doing what they want to do, we know what's best & we've marked it as unassigned in jira (still bad)
- "assign" in parabol mean "assign" in jira (this doesn't follow Agile, but is safe & predictable)
- you can only publish tasks that are assigned to you (feels half-baked)
- same as above, but when you click publish, the assignee becomes "unassigned", then when you set it again, it assigns it in both systems (again, we're doing something that the user didn't ask for because we know better)

I don't see the harm in adding an "assignee" when creating a jira issue. I do see harm in ignoring a user-defined "assignee" value in our own software.

**Comment by jordanh** [comment]
Discussed via meeting, resolved to proceed with @mattkrick's proposed AC:

- The task creator is the reporter. If the task creator is not on Jira, then the task assignee is the reporter.
- The assignee is the assignee

**Comment by ackernaut** [comment]
@mattkrick I think the UI components for the card menus are less of a mystery, do I need to create views in Figma for the integration grid view?

**Comment by mattkrick** [comment]
@ackernaut yes please! Could you do a quick design of everything under the `INTEGRATION GRID VIEW` heading in the first comment?
